### PR TITLE
Maybe a better way to match incomming chat. 

### PR DESCRIPTION
**Pull request by kepet19**

### CODE REVIEW

**Comment by terrarier2111** [comment]
This is so cool, great work!

**Comment by kepet19** [comment]
I just found out that there is translate of Block's aswell. 
I am going to rethink this. maybe just let the translate be a string instead of and Enum? 
We already got the name of the block in hoverEvent, but I am not sure how the older server do yet. 


`Chat`
```json 
{
  "italic": true,
  "color": "gray",
  "translate": "chat.type.admin",
  "with": [
    { "text": "Server"  },
    { "translate": "commands.give.success.single",  "with": [
        "300",
        { "color": "white",
          "hoverEvent": { "action": "show_item", "contents": {
                                                                    "id": "minecraft:dirt",
                                                                    "count": 300
                                                                    }
          },
          "translate": "chat.square_brackets",
          "with": [
            { "extra": [
                { "translate": "block.minecraft.dirt"   }
              ],
              "text": ""
            }
          ]
        },
        {
          "insertion": "Kevz__",
          "clickEvent": {
            "action": "suggest_command",
            "value": "/tell Kevz__ "
          },
          "hoverEvent": {
            "action": "show_entity",
            "contents": {
              "type": "minecraft:player",
              "id": "b4309f1c-6a41-4ac0-8857-2be00133108d",
              "name": {
                "text": "Kevz__"
              }
            }
          },
          "text": "Kevz__"
        }
      ]
    }
  ]
}

```

**Comment by terrarier2111** [comment]
For what do you need this void dependency?

**Comment by kepet19** [comment]
> For what do you need this void dependency?

Nothing. 
it is a create for meaning nothing, it was for error's I don't need. Whoops

Emm I think I have found a better way. 
it will take some time, but it should be a better parser than what I have done already, more simple, also be more scalable. fingers cross.

**Comment by terrarier2111** [comment]
> > For what do you need this void dependency?
> 
> Nothing. it is a create for meaning nothing, it was for error's I don't need. Whoops
> 
> Emm I think I have found a better way. it will take some time, but it should be a better parser than what I have done already, more simple, also be more scalable. fingers cross.

That sounds extremely promising

**Comment by kepet19** [comment]
> > There is missing a lot of block and items translations. maybe we should add that as an issue that is at least 500 lines more, not sure where to get the data.
> > Most of the commands is translated.
> 
> What do you mean with the "that" in "maybe we should add that as an issue that is at least 500 lines more"

Rest of the translation of these types, should have an issues for it self, because it is a pretty big task.
```rust
"block.minecraft.dirt" => "Dirt",
"item.minecraft.diamond" => "Diamond",
```
I want to finish this pr first and then I would look at the rest of the translations.

File: protocol/src/format.rs [hunk 1]
```diff
@@ -12,22 +12,22 @@
...
+pub use crate::format::color::*;
+
 #[derive(Debug, Clone)]
 pub enum Component {
     Text(TextComponent),
 }
 
 impl Component {
     pub fn from_string(str: &str) -> Self {
-        // TODO: Support: §l, §k, §m, §o, §n, §r
```
Comment (terrarier2111) [inline protocol/src/format.rs:25]: Did you already add support for all of these modifiers or why did you remove this comment?

Comment (kepet19) [reply protocol/src/format.rs:25]: > Did you already add support for all of these modifiers or why did you remove this comment?

Maybe I should add the comment back or phase if differently.
because the modifiers get set. it just don't get rendered.
or maybe I should add the comment to where the modifiers get read?

Comment (terrarier2111) [reply protocol/src/format.rs:25]: I think you should add it to where they get added or where they are stored and rephrase it a bit

File: protocol/src/format.rs [hunk 2]
```diff
@@ -12,89 +12,230 @@
...
+use serde::Deserialize;
+
+pub use crate::format::color::*;
+
+const LEGACY_CHAR: char = '§';
 
 #[derive(Debug, Clone)]
-pub enum Component {
-    Text(TextComponent),
+pub struct Component {
```
Comment (terrarier2111) [inline protocol/src/format.rs:27]: Are you sure, it's correct to transform Component into a struct? Because how did you plan to support clickable/hoverable/web link components?

Comment (kepet19) [reply protocol/src/format.rs:27]: I would make the rest ready for review tomorrow.

File: src/ui/mod.rs [hunk 3]
```diff
@@ -1248,18 +1248,21 @@ impl<'a> ElementHolder for FormatState<'a> {
...
+            self.append_text(
+                self.scale_x,
+                self.scale_y,
+                componet.get_text(),
+                componet.get_modifier().color,
+            )
+        }
+    }
+
+    fn build_with_optional_color(&mut self, componets: &format::Component, color: format::Color) {
```
Comment (terrarier2111) [inline src/ui/mod.rs:1262]: Here is a typo: componets

File: src/ui/mod.rs [hunk 4]
```diff
@@ -1248,18 +1248,21 @@ impl<'a> ElementHolder for FormatState<'a> {
...
+        for componet in componets.list.iter() {
```
Comment (terrarier2111) [inline src/ui/mod.rs:1263]: Same here for: componet

File: src/entity/player.rs [hunk 5]
```diff
@@ -453,9 +453,11 @@ impl CustomEntityRenderer for PlayerRenderer {
                 y_scale: 0.16,
                 x_scale: 0.01,
             };
-            let mut name = format::Component::Text(format::TextComponent::new(&player_model.name));
-            format::convert_legacy(&mut name);
-            state.build(&name, format::Color::Black);
+            let name = format::Component::new(format::ComponentType::new_with_color(
```
Comment (terrarier2111) [inline src/entity/player.rs:456]: new_with_color sounds a bit odd, what about with_color or adding an optional to the new function

File: src/entity/block_entity/sign.rs [hunk 6]
```diff
@@ -179,7 +179,9 @@ impl ecs::System for SignRenderer {
                 y_scale: Y_SCALE,
                 x_scale: X_SCALE,
             };
-            state.build(line, format::Color::Black);
+            // TODO: make sure this is correct!
+            // state.build(line, format::Color::Black);
+            state.build_with_optional_color(line, format::Color::Black);
```
Comment (terrarier2111) [inline src/entity/block_entity/sign.rs:184]: The naming seems a bit odd here as well, how about adding an Option to the build function

File: protocol/src/format.rs [hunk 7]
```diff
@@ -104,322 +253,444 @@ impl Component {
...
+                    green: 0x34,
+                    blue: 0x56,
                 }
-                txt.text = "".to_owned();
-            }
+            ),
+            _ => panic!("Could not parse hex color correct"),
+        }
+
+        match Color::from_str("rEd") {
```
Comment (terrarier2111) [inline protocol/src/format.rs:683]: The naming in these from_strs seems to be inconsistent

File: protocol/src/translate.rs [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+            dbg.field("extra", &self.extra);
+        }
+        if self.with.len() > 0 {
+            dbg.field("with", &self.with);
+        }
+        dbg.finish_non_exhaustive()
+    }
+}
+
+pub fn translate(str: &str) -> &str {
```
Comment (terrarier2111) [inline protocol/src/translate.rs:109]: I think you should add a comment about adding support for multiple languages

**Review by terrarier2111 — CHANGES_REQUESTED** [review]

File: protocol/src/format.rs (hunk 7 above)
Comment (terrarier2111) [reply protocol/src/format.rs:683]: no, i mean like the color in th apostrophes, like it is sometimes in caps and sometime not

Comment (kepet19) [reply protocol/src/format.rs:683]: Ahh the test I change that.

File: protocol/src/format.rs [hunk 9]
```diff
@@ -369,20 +369,20 @@ pub enum ComponentType {
...
-
-    pub fn new_with_color(val: &str, color: Color) -> Self {
-        Self::Text {
-            text: val.to_owned(),
-            modifier: Modifier {
-                color,
-                ..Default::default()
-            },
+    pub fn new(val: &str, color: Option<Color>) -> Self {
+        if let Some(color) = color {
```
Comment (terrarier2111) [inline protocol/src/format.rs:373]: This can be simplefied, as much of the code is duplicated here, as it just diverges in the modifier field value

**Review by terrarier2111 — CHANGES_REQUESTED** [review]

File: protocol/src/format.rs (hunk 9 above)
Comment (kepet19) [reply protocol/src/format.rs:373]: You are thinking like this ? 

```rust
    pub fn new(val: &str, color: Option<Color>) -> Self {
        Self::Text {
            text: val.to_string(),
            modifier: match color {
                Some(color) => Modifier { color, .. },
                None => Modifier::default(),
            },
        }
    }
```

Comment (terrarier2111) [reply protocol/src/format.rs]: exactly!

**Review by terrarier2111 — APPROVED** [review]
